src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java

```diff
@@ -195,32 +208,59 @@ public void updateWithReaction(Post post, Reaction reaction, Long courseId) {
-     * @param exerciseId        id of the exercise for which the posts should be fetched
-     * @param lectureId         id of the lecture for which the posts should be fetched
-     * @return list of posts that match the given context
+     * @param pagingEnabled     fetches single page instead of all entities
+     * @param pageable          requested page and page size
+     * @param postContextFilter request object to fetch posts
+     * @return page of posts that match the given context
      */
-    public List<Post> getPostsInCourse(Long courseId, CourseWideContext courseWideContext, Long exerciseId, Long lectureId) {
+    public Page<Post> getPostsInCourse(boolean pagingEnabled, Pageable pageable, @Valid PostContextFilter postContextFilter) {
+
+        List<Post> postsInCourse;
         // no filter -> get all posts in course
-        if (courseWideContext == null && exerciseId == null && lectureId == null) {
-            return this.getAllCoursePosts(courseId);
+        if (postContextFilter.getCourseWideContext() == null && postContextFilter.getExerciseId() == null && postContextFilter.getLectureId() == null) {
+            postsInCourse = this.getAllCoursePosts(postContextFilter);
         }
         // filter by course-wide context
-        else if (courseWideContext != null && exerciseId == null && lectureId == null) {
-            return this.getAllPostsByCourseWideContext(courseId, courseWideContext);
+        else if (postContextFilter.getCourseWideContext() != null && postContextFilter.getExerciseId() == null && postContextFilter.getLectureId() == null) {
+            postsInCourse = this.getAllPostsByCourseWideContext(postContextFilter);
         }
         // filter by exercise
-        else if (courseWideContext == null && exerciseId != null && lectureId == null) {
-            return this.getAllExercisePosts(courseId, exerciseId);
+        else if (postContextFilter.getCourseWideContext() == null && postContextFilter.getExerciseId() != null && postContextFilter.getLectureId() == null) {
+            postsInCourse = this.getAllExercisePosts(postContextFilter);
         }
         // filter by lecture
-        else if (courseWideContext == null && exerciseId == null && lectureId != null) {
-            return this.getAllLecturePosts(courseId, lectureId);
+        else if (postContextFilter.getCourseWideContext() == null && postContextFilter.getExerciseId() == null && postContextFilter.getLectureId() != null) {
+            postsInCourse = this.getAllLecturePosts(postContextFilter);
         }
         else {
             throw new BadRequestAlertException("A new post cannot be associated with more than one context", METIS_POST_ENTITY_NAME, "ambiguousContext");
         }
+
+        // search by text or #post
+        if (postContextFilter.getSearchText() != null) {
+            postsInCourse = postsInCourse.stream().filter(post -> postFilter(post, postContextFilter.getSearchText())).collect(Collectors.toList());
```

Comment:
```suggestion
            postsInCourse = postsInCourse.stream().filter(post -> postFilter(post, postContextFilter.getSearchText())).toList();
```

Reply:
I started getting a null pointer exception after this change. The behavior of the code above is different. As far as I understood, using .toList(); makes the list uneditable which I must do after this code. You can check the following failing test outputs; https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS3050-JAVATEST-32/test/case/491283038

Reply:
Thanks, I revoked this part. You can kindly take a look now. @b-fein